bucket/leaky_bucket.go

```diff
@@ -40,61 +49,37 @@ func (b *LeakyBucket) StopDripping() {
 	}()
 }
 
-func (b *LeakyBucket) UnsafeAvailable() int {
-	return b.available
+func (b *LeakyBucket) Available() int {
+	return len(b.bucket)
 }
 
-func (b *LeakyBucket) IsEmpty() (success bool) {
-	result := b.locked(func() interface{} {
-		return b.available <= 0
-	})
-	success, _ = result.(bool)
-	return
+func (b *LeakyBucket) IsEmpty() bool {
+	return len(b.bucket) == 0
 }
 
-func (b *LeakyBucket) IsFull() (success bool) {
-	result := b.locked(func() interface{} {
-		return b.available == b.size
-	})
-	success, _ = result.(bool)
-	return
+func (b *LeakyBucket) IsFull() bool {
+	return len(b.bucket) >= cap(b.bucket)
 }
 
 func (b *LeakyBucket) TopUp() {
-	b.fillBucket(true)
+	b.fill(b.Size)
 }
 
 func (b *LeakyBucket) AddDrops() {
-	time.Sleep(time.Duration(b.duration) * time.Second)
-	b.fillBucket(false)
+	b.fill(b.Refill)
 }
 
 func (b *LeakyBucket) GetDrop() {
+	log.Println("Getting drop ", len(b.bucket), "available")
 	<-b.bucket
-	b.locked(func() interface{} {
-		b.available--
-		return nil
-	})
 }
 
-type lockedFunc func() interface{}
-
-func (b *LeakyBucket) locked(code lockedFunc) interface{} {
-	b.lock.Lock()
-	defer b.lock.Unlock()
-	return code()
-}
-
-func (b *LeakyBucket) fillBucket(full bool) {
-	amount := b.size - b.available
-	if !full && amount > b.refill {
-		amount = b.refill
-	}
+func (b *LeakyBucket) fill(amount int) {
 	for i := 0; i < amount; i++ {
-		b.bucket <- true
-		b.locked(func() interface{} {
-			b.available++
-			return nil
-		})
+		select {
```

Comment:
This is really cool and what makes this work without locks. You can use selects not only on reading from channels but also for writing to them!